### PR TITLE
feat(routing): implement FloatConverter for #2022

### LINKED ISSUE #2022: Implement FloatConverter
Implement `FloatConverter` along the lines of [`IntConverter`](https://falcon.readthedocs.io/en/stable/api/routing.html#falcon.routing.IntConverter). Draw inspiration from `IntConverter`, or even find an efficient way to share code between the two!

Add the new converter to the list of [Built-in Converters](https://falcon.readthedocs.io/en/stable/api/routing.html#built-in-converters) under the `float` identifier.

Open questions: should we support converting `nan`, `inf` & `-inf` from path?

### PR DESCRIPTION
**Pull request by meetshah133**
Closes #2022

Changes: 

Modified the existing logic of IntConverter class to reuse it in FloatConverter.
Added he new converter to the list of [Built-in Converters](https://falcon.readthedocs.io/en/stable/api/routing.html#built-in-converters) under the float identifier.

TODO address before merging:

- [x] Clean up code formatting and make the `pep8` and `black` Tox environments pass.
- [x] Write missing tests for `FloatConverter`. We require 100% branch coverage at all times.
- [x] Add `2022.newandimproved.rst` Towncrier [newsfragment](https://github.com/falconry/falcon/tree/master/docs/_newsfragments).
- [x] Find a suitable replacement for `num_digits`. Or maybe we can remove it altogether?
- [x] Do not use `camelCase` unless there is a compelling reason.

### CODE REVIEW

**Comment by meetshah133** [comment]
Hi @vytas7, can you please help by reviewing the code changes ?
I have tested the FloatConverter and IntConverter class they seems working fine,
Will be happy to get more inputs from you and contribute to the project.

**Comment by vytas7** [comment]
Hi @meetshah133, and thanks for this PR!
Unfortunately I won't have the bandwidth for this today, but will try to look at your code it in the beginning of the next week. 
It also possible that @CaselIT, @kgriffs or other maintainers beat me to it :calendar:

To start with, you could address the CI failures (see **"Some checks were not successful"** below) to get the things moving faster :arrow_down:
(It's lame that GitHub doesn't run Actions for first-time contributors, but I can try to at least push the button when I notice that you have committed new stuff.)

File: falcon/routing/converters.py
Comment (CaselIT) [inline falcon/routing/converters.py:115]: slots is not needed, since it will inherit the `IntConverter` ones

**Review by CaselIT — CHANGES_REQUESTED** [review]
I'm not sure if `num_digits` makes sense for a float, otherwise it seems mostly ok.

The test are missing. You found the existing ones in test_uri_converters.py

File: falcon/routing/converters.py
Comment (myusko) [inline falcon/routing/converters.py:69]: I'd suggest using `snake_case` instead of `camelCase`, in order to follow the best Python practices and recommendations.

See: https://www.python.org/dev/peps/pep-0008/#function-and-variable-names

File: falcon/routing/converters.py
Comment (vytas7) [inline falcon/routing/converters.py:107]: Indeed, as @CaselIT noted, `num_digits` doesn't really make sense for `float`s.
What is the number of digits in `2.9999999999999(9)`? `inf`? `1.23456789e123`?

We could either remove this parameter, or maybe introduce a different one? Like `significant_digits`? And then just round to the specified precision instead of failing validation.

**Review by vytas7 — CHANGES_REQUESTED** [review]
Adding to what @CaselIT and @myusko pointed out, we should also add a `towncrier` newsfragment for this new feature.

(I've edited your first comment and added a checklist of what's remaining.)

**Comment by meetshah133** [comment]
Thanks @vytas7 , @CaselIT  , @myusko for your valuable inputs.
I will work on those items suggested by you all  and submit a new commit to this repository.

**Comment by meetshah133** [comment]
Hi @vytas7 , have implemented all the suggestions, need some guidance on "tox -e black" test failure

**Comment by vytas7** [comment]
Luckily, that's the easiest one ;) Just install the latest `black` version into your environment, and run it:
```
$ pip install -U black
$ black .
```

**Comment by meetshah133** [comment]
Thanks @vytas7 finally all checks passed

File: falcon/routing/converters.py
Comment (CaselIT) [inline falcon/routing/converters.py:109]: Since this class extends int converter that already defines min and max slots should be empty in the subclass, since no new attribute is defined

File: falcon/routing/converters.py
Comment (CaselIT) [inline falcon/routing/converters.py:131]: This seems like it's not needed

File: tests/test_uri_converters.py
Comment (CaselIT) [inline tests/test_uri_converters.py:80]: I think we could add some additional cases:
- Negative values
- float values (we are only passing ints)
- inf, -inf and nan

**Review by CaselIT — CHANGES_REQUESTED** [review]
Overall it looks ok, thanks.

I've just listed some little changes

**Comment by meetshah133** [comment]
@CaselIT and @vytas7  thanks for your inputs, I have created a new commit based on your inputs.

**Review by CaselIT — DISMISSED** [review]

File: falcon/routing/converters.py
Comment (vytas7) [inline falcon/routing/converters.py:91]: Maybe this method could be made private? Like `_validate_min_mix(value)`? Unless we want to document it and make it official (I don't think so).

**Review by vytas7 — CHANGES_REQUESTED** [review]
It's shaping up :+1: I'd still like to sand off some rough edges before merging this though (provided feedback inline).

**Comment by vytas7** [comment]
Hi again @meetshah133, just checking if you are still working on these changes?

**Comment by meetshah133** [comment]
> Hi again @meetshah133, just checking if you are still working on these changes?

Hi @vytas7, was not well for past few days.
I am working on the changes suggested by you will soon commit the changes.
Thanks.

**Comment by vytas7** [comment]
Awesome, thanks for the update!

File: falcon/routing/converters.py
Comment (vytas7) [reply falcon/routing/converters.py:109]: The parameter could also be called `allow_nan` in line with the stdlib's [`json.dump`](https://docs.python.org/3/library/json.html#json.dump), just it should default to `False` in our case. IMHO.

**Comment by meetshah133** [comment]
Hi @vytas7 , I have done the commit there are 3 test failing, but I am unable to understand why.
Could you please assist ?

File: falcon/routing/converters.py
Comment (CaselIT) [inline falcon/routing/converters.py:103]: allow_nan is missing

File: falcon/routing/converters.py
Comment (CaselIT) [inline falcon/routing/converters.py:115]: I believe this check needs be after `value = float(value)` otherwise here it's comparing a string, so `value != value` is always false

File: tests/test_uri_converters.py
Comment (CaselIT) [inline tests/test_uri_converters.py:110]: I think test with nan are missing

**Review by CaselIT — CHANGES_REQUESTED** [review]
looks mostly good, just a couple of points

**Comment by vytas7** [comment]
@meetshah133 could you please update your branch with the latest changes from `master` instead of force-pushing over?

**Comment by kgriffs** [comment]
OK folks, in the interest of getting this over the finish line sooner rather than later, I took the liberty of trying to address the unresolved feedback. Please take another look. Thanks!

**Review by CaselIT — DISMISSED** [review]
just a couple of minor things, but it's fine either way for me

File: falcon/routing/converters.py
Comment (vytas7) [reply falcon/routing/converters.py:143]: I think it actually was a misunderstanding overall wrt my initial suggestion, maybe partially because naming is hard...
IMHO it would make sense to have a single parameter governing whether we allow other than finite ("ordinary") numbers or not. So basically `nan`, `inf`, `-inf`.

Comment (CaselIT) [reply falcon/routing/converters.py:143]: makes sense. so actually use `math.isfinite(value)`

Comment (kgriffs) [reply falcon/routing/converters.py:143]: The inequality trick executes a tiny bit faster vs invoking `math.isnan()`. But that's a moot point if we follow @vytas7's suggestion. I can push a commit to make that change.

File: falcon/routing/converters.py
Comment (kgriffs) [reply falcon/routing/converters.py:136]: good catch. I'll address this one as well.

File: falcon/routing/converters.py
Comment (kgriffs) [reply falcon/routing/converters.py:147]: done

Comment (kgriffs) [reply falcon/routing/converters.py:147]: let me know what you think

**Review by vytas7 — APPROVED** [review]
LGTM now :+1:

**Review by CaselIT — APPROVED** [review]

**Pull request merged**
